### PR TITLE
Add test for casting in `test_nsgaii.py`

### LINKED ISSUE #3982: unrobust optuna sampling behaviour
### Expected behavior

I have something like this in the code: `trial.suggest_categorical('somename', [False, True])`

and when it is running sometimes it gives an error like `ValueError: 'True' not in (False, True).`, but sometimes it samples just fine without this kind of error.

I looked into issue with pdb and at `optuna/samplers/nsgaii/_crossover.py(115)` you actually have a `child_params` dictionary which contains  sampled parameters and there is `{..., 'somename': "False"}`, although it would make much better sense to have that parameter sampled in `boolean` values.

Any ideas why that would be happening? Thank you in advance.

### Environment

- Optuna version: 3.0.1
- Python version: 3.8.13
- OS:Linux-4.15.0-1051-aws-x86_64-with-glibc2.17
- (Optional) Other libraries and their versions:

### Error messages, stack traces, or logs

```shell
File "/lib/python3.8/site-packages/optuna/study/study.py", line 419, in optimize
    _optimize(
  File "/lib/python3.8/site-packages/optuna/study/_optimize.py", line 66, in _optimize
    _optimize_sequential(
  File "/lib/python3.8/site-packages/optuna/study/_optimize.py", line 160, in _optimize_sequential
    frozen_trial = _run_trial(study, func, catch)
  File "/lib/python3.8/site-packages/optuna/study/_optimize.py", line 187, in _run_trial
    trial = study.ask()
  File "/lib/python3.8/site-packages/optuna/study/study.py", line 514, in ask
    trial = trial_module.Trial(self, trial_id)
  File "/lib/python3.8/site-packages/optuna/trial/_trial.py", line 54, in __init__
    self._init_relative_params()
  File "/lib/python3.8/site-packages/optuna/trial/_trial.py", line 63, in _init_relative_params
    self.relative_params = self.study.sampler.sample_relative(
  File "/lib/python3.8/site-packages/optuna/samplers/nsgaii/_sampler.py", line 211, in sample_relative
    child_params = perform_crossover(
  File "/lib/python3.8/site-packages/optuna/samplers/nsgaii/_crossover.py", line 115, in perform_crossover
    if _is_contained(child_params, search_space):
  File "/lib/python3.8/site-packages/optuna/samplers/nsgaii/_crossover.py", line 160, in _is_contained
    if not param_distribution._contains(param_distribution.to_internal_repr(param)):
  File "/lib/python3.8/site-packages/optuna/distributions.py", line 516, in to_internal_repr
    raise ValueError(f"'{param_value_in_external_repr}' not in {self.choices}.")
ValueError: 'True' not in (False, True).
```

### Steps to reproduce

1.
2.
3.
```python
# python code
```

### Additional context (optional)

_No response_

### PR DESCRIPTION
**Pull request by Alnusjaponica**
## Motivation
Resolve #3982
## Description of the changes
Added test to confirm that type of categorical variable is not unexpectedly casted in `NSGAIISampler`.
## Alternatives
Test `_crossover.perform_crossover` or `_crossover._try_crossover` directly.

### CODE REVIEW

**Comment by Alnusjaponica** [comment]
@HideakiImamura I made this PR review ready. Could you assign me reviewers?

**Comment by toshihikoyanase** [comment]
@nzw0301 Could you review this PR as an author of #4035, please?
@HideakiImamura You're familiar with the crossover implementation, so could you review this PR, please?

**Review by nzw0301 — APPROVED** [review]
Thanks! I confirmed the test with optuna 3.0.1 fails.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I think @toshihikoyanase's comment is reasonable. Except that comment, LGTM.

File: tests/samplers_tests/test_nsgaii.py
Comment (Alnusjaponica) [reply tests/samplers_tests/test_nsgaii.py]: Thank you for your indication.
I think using two different types of categorical_values was essential for the error (https://github.com/optuna/optuna/issues/3982#issuecomment-1263194047), but I agree that it will be more simple to write test cases without using combination. I fixed it accordingly.

Comment (toshihikoyanase) [reply tests/samplers_tests/test_nsgaii.py]: Hmm, the new version of the test function has more lines than before.

> I think using two different types of categorical_values was essential for the error (https://github.com/optuna/optuna/issues/3982#issuecomment-1263194047)

The minimum requirement of the test case seems to be a combination of raw values and stringified values.
If so, how about just parametrizing raw values? I guess stringified values can be generated in the test case.
The following is just an example.

```python
@pytest.mark.parametrize("choices", [[-1, 0, 1], [True, False], ["string1", "string2"]])
def test_crossover_casting(choices: List[Any]) -> None:
    def objective(trial: optuna.Trial) -> Sequence[float]:
        str_choices = list(map(str, choices))
        r = trial.suggest_categorical("r", choices)
        s = trial.suggest_categorical("s", str_choices)
        assert isinstance(r, type(choices[0]))
        assert isinstance(s, type(str_choices[0]))
        return 1.0, 2.0
```

This test case lacks the pair of numbers and bools, but I don't think it is critical because the bug just requires two different types of values (e.g., numbers and strings, or bools and strings).

Comment (toshihikoyanase) [reply tests/samplers_tests/test_nsgaii.py]: Thank you for applying the suggestion to the PR.
@HideakiImamura What do you think of this simplification?

**Review by HideakiImamura — APPROVED** [review]
LGTM.

File: tests/samplers_tests/test_nsgaii.py
Comment (HideakiImamura) [reply tests/samplers_tests/test_nsgaii.py]: Looks very good.

**Comment by HideakiImamura** [comment]
@toshihikoyanase Please merge this PR after your approval.

**Review by toshihikoyanase — APPROVED** [review]
LGTM! Thank you.

**Pull request merged**
